### ISSUE TITLE
How to add a non-param type similar to `sourceVersion` 

### ISSUE THREAD

**Issue by arezaii** [chapel-lang/chapel#19201]
### Overview

During the module review of the `Version` module we discussed a need for `mason` to support a type that describes a software version. 
`mason` needs this type to reason about different dependency versions. These versions are never known at compile time, so the current `sourceVersion` type doesn't meet the needs and [`mason` implements a custom type](https://github.com/chapel-lang/chapel/blob/f596c4cf35fb4226b69798a0bcd9f0fb5b2d3127/tools/mason/MasonUtils.chpl#L227) which replicates and builds on the functionality found in the `sourceVersion` type.

A draft PR is up at https://github.com/chapel-lang/chapel/pull/19300, here are the bullet points from the proposal: 

- Add a new `programVersion` record whose fields are `var` rather than `param`
- Maintain a similar API from `sourceVersion` with some exceptions listed below
  - Use a new `createMutableVersion` helper to avoid accidentally creating a `sourceVersion` instead of a `programVersion`. This could just as easily be named `createProgramVersion`
  - don't print "version " when casting a `programVersion` to string. The rationale is that as a user, I just want the representation of the object as a string and adding words to it means I'd have to strip them back out rather than add them to my output message as needed. For example, what if my program should format the output as `v2.1.2`? 

### Example Usage

```chapel
use Version;

var v2_1_0 = createMutableVersion(2,1,0);
var v2_1_1_xyz = createMutableVersion(major=2, minor=1, update=1, commit="xyz");
var v3_1_2 = new programVersion(major=3, minor=1, update=2);

```

### Suitability for `mason`

To ensure the new `programVersion` could replace the custom `VersionInfo` in `mason`, I have a branch that demonstrates the changeover at [https://github.com/arezaii/chapel/tree/mason-programVersion-refactor](https://github.com/chapel-lang/chapel/compare/main...arezaii:mason-programVersion-refactor)

### Feedback requests

- Are we good with proceeding in this general direction?
- Are we happy with the `programVersion` type name?
- Do we really want or need a helper that is synonymous with `new programVersion`?
  - If so, are we happy with the `createMutableVersion` helper name and argument names?
- Is the proposed cast-to-string behavior appropriate?
- Is it OK that this is a record rather than a class?
- Anything else?

**Comment by arezaii** [comment]
Working through this implementation, we may be better with two types. I think if we try to make a param and non-param version of a single type we will lose the meaning of having a param version, as any use could be valid by choosing the non-param initializer. I put a draft PR with a possible implementation here: https://github.com/chapel-lang/chapel/pull/19300

**Comment by lydia-duncan** [comment]
I think I prefer `createProgramVersion` to `createMutableVersion` but am otherwise happy with this direction

**Comment by bradcray** [comment]
> Are we good with proceeding in this general direction?

I think so.

> Are we happy with the programVersion type name?

I feel so-so about it, to a large extent because it feels hard to distinguish between `sourceVersion`.  I think `sourceVersion` is arguably justifiable as a name because it can be embedded into the source code of a program, and so known statically and unchanging.  But I tend to think of source and program as being very related, nearly interchangeable terms, and I'm not sure that they're distinct enough to clearly say "This is param / known at compile-time" vs. "This is non-param / only known at execution time."  That said, there may be a mnemonic or rationale that just hasn't occurred to me yet, _and_ I don't have a better suggestion in mind yet.

> Do we really want or need a helper that is synonymous with new programVersion?

That's a good question, and I think the same could/should be asked for the static/param case as well.  IIRC, we may have originally just used `version` for this so that you could say `if chplVersion < version(1.23)` but upon making it more verbose, the benefit seems reduced to me.  I think @ronawho was an original champion of the `version()` form, so may have more thoughts here.

In any case, I think we should support standalone factories for both version types or neither.  As stated elsewhere (but without generating support from others), I generally find `createT()` to generally be a mouthful compared to `newT()` or `new T()`).

> If so, are we happy with the createMutableVersion helper name

Logically, once we have two version types, ideally the method names would clearly reflect the two cases (where my intuition here would either be to call this `createProgramVersion` or to rename the type to `mutableVersion`).

> and argument names?

It'd be good to put the argument names into the OP or question, but checking the PR, I see:

```
major: int,
                     minor: int,
                     update: int = 0,
                     commit: string = ""
```

which seems plenty clear to me.

> It the proposed cast-to-string behavior appropriate?

It seems to me that the casts for both types should behave similarly, so maybe this should be forked off into its own issue w.r.t. the existing capability, or part of the next module re-review?

> Is it OK that this is a record rather than a class?

Yeah, I think records are preferable for this sort of thing.

**Comment by arezaii** [comment]
Thanks for providing feedback! I have made some changes to the OP based on your comments and addressed the other responses below.


> I feel so-so about it, to a large extent because it feels hard to distinguish between sourceVersion. 

I considered `runtimeVersion` as an alternative, but didn't see too much to distinguish it from `programVersion`. I don't have any attachment to the current names and would be happy to evaluate alternatives.

> In any case, I think we should support standalone factories for both version types or neither. As stated elsewhere (but without generating support from others), I generally find createT() to generally be a mouthful compared to newT() or new T()).

Agreed. I also don't _love_ `createT()`, but also dislike `newT()` because verbally, it's indistinguishable from `new T()` (complicates pair programming). I just don't know what the factory function gets us in this case because it doesn't really save any typing, and the arguments are exactly the same. Maybe it's just for alignment across our included types?

> Logically, once we have two version types, ideally the method names would clearly reflect the two cases (where my intuition here would either be to call this createProgramVersion or to rename the type to mutableVersion).

Agreed

> It'd be good to put the argument names into the OP or question

I updated the example to show the names.

> >Is the proposed cast-to-string behavior appropriate?

> It seems to me that the casts for both types should behave similarly, so maybe this should be forked off into its own issue w.r.t. the existing capability, or part of the next module re-review?

Yes, I thought about this after writing out the rationale for dropping it, and it seems like dropping it in both places makes sense for the same reason, but I'll spawn another issue for that.

**Comment by bradcray** [comment]
> I just don't know what the factory function gets us in this case because it doesn't really save any typing, and the arguments are exactly the same

I agree it's not buying us a ton.  When it was just called `version()` I think it was easy to argue that it improved readability.  I can't recall what arguments were used to change it to `createVersion()` rather than just leaning on `new sourceVersion()`, though I expect that it related to the fact that it reads more like English and avoids needing to know that there is a type called `sourceVersion`.  But that worked better when there was just one version type, less well with two.

**Comment by mppf** [comment]
Re the name `programVersion`, `runtimeVersion` sounds better to me, but I would think it is referring to the Chapel C runtime, so maybe `runningVersion` ? But that doesn't seem great in the mason use case. Maybe `builtVersion` or `artifactVersion` or `softwareVersion` or `packageVersion` ?

**Comment by arezaii** [comment]
Along the lines of `builtVersion`, I also considered `compiledVersion` to help differentiate from `sourceVersion`. 
I do like `packageVersion` as well.

**Comment by arezaii** [comment]
I have been mulling this over for a while now and I just don't feel like there's a great name to distinguish what we call `sourceVersion` from one that can be built at runtime. How do people feel about possibly renaming `sourceVersion`? One pair of names that I thought might make a good distinction is `staticVersion` vs. `dynamicVersion`. Any thoughts or concerns about that idea? Is there too much C baggage tied to `static`?

**Comment by bradcray** [comment]
As a compiler and C person, the static vs. dynamic distinction definitely makes sense to me, but we have had users turn up their noses at `static` in other, past contexts for being too jargon-y / non-intuitive, which is part of the reason we don't use it as a keyword (e.g., `static const` vs. `param`).  Even `dynamic` in terms of English meaning, suggests things to me like "likely to be changing over time" as opposed to simply "not known at compile time".

This naming challenge sortof reminds me of the clunkiness around the names `compilerError()` and `compilerWarning()`.  E.g., `compileTimeVersion` seems very precise, but wordy/clunky.  `paramVersion` would be consistent with `param`, but not being much of a fan of `param` itself as a keyword, I'm not super-excited about it either.

Generally, though, I'm open to renaming `sourceVersion` if we can find the right terms to rename to.

This discussion also brings up a longstanding desire to have some way of param-choosing between `param` and `const` with a single type, such that `param(true) x: int;` would result in a param and `param(false) x: int;` would result in a `const`.  If we had this, we could (potentially?) represent both cases with a single type, like so:

```chapel
record version {
  param compileTime: bool;
  param(compileTime) major, minor, update: int;
}
```

such that `version(compileTime=true)` would be the current static version and `version(compileTime=false)` would be the dynamic one.

**Comment by mppf** [comment]
> `paramVersion` would be consistent with `param`, but not being much of a fan of `param` itself as a keyword, I'm not super-excited about it either.

`paramVersion` and `version` seem "obvious" to me. For anybody who wants to remember the discussion about renaming `param`, it is in #5066. IMO if we are not going to rename `param` then we should be happy with `paramVersion`. (I am personally open to revisiting that `param` naming discussion).

In terms of totally different directions, is there some way we can allow a `param` record, where the use of the record indicates it is param or not, just like the use of an `int` does? I.e. `param x: sourceVersion` would be a compile-time version and `var x: sourceVersion` would be a run-time one.

I think that the immediate language problem that appears with that is - supposing we had such a record - how do we declare the fields? Perhaps

``` chapel
record version {
  const major, minor, update: int;
}
```

is good enough, but it would bring a different meaning to `const` (since now, when the whole record is `param`, the `const` is also `param`). Arguably we would need some way to introduce the fields that meant "const or param". One proposal for that is `param?` (which I think has been proposed in argument passing contexts). Or, we could use Brad's proposal and have the compiler fill in the `compileTime` field.

**Comment by arezaii** [comment]
If I understand correctly, you both (@mppf and @bradcray) are talking about options that require updates to the language in order to be implemented. If that's correct, I'm not sure how to proceed with changes to the `Version` module at the moment. I can continue to suggest new names, or we can make this issue dependent on the resolution of these bigger language design questions/changes, or perhaps a third option I'm not seeing.

**Comment by bradcray** [comment]
Here's my sketch of @benharsh's suggestion to unify these types today, which I haven't tried to compile but seems like an approach that may work with only minor tweaks/fixes to me (and matches what I think I heard others saying as well):

```chapel
record version {
  param flag: bool;  // is this a param or non-param thing?  This needs a better name.

  param pmajor, pminor, pupdate = -1;  // These are terrible names too; really, they want to be private, but we don't have that yet... :'(
  // think we need a sha string here, but I'm not looking it up to remind myself currently

  const cmajor, cminor, cupdate: if bool then int else void;
  // ditto on sha

  proc major param where flag == true {
    return pmajor;
  }
  
  proc major where flag == false {
    return cmajor;
  }

  // and same for the other fields
}

proc createVersion(param major: int, param minor: int, param update: int, ...) {
  return new version(flag=true, pmajor=major, pminor=minor, pupdate=update);
}

proc createVersion(major: int, minor: int, update: int, ...) {
  return new version(flag=false, cmajor=major, cminor=minor, cupdate=update);
}
```

Specifically, note that I chose not to try and make the params `void` if flag is `false` because I was worried our generics machinery wouldn't be up to the task (i.e., my overall generic type depends on the value of earlier generic fields).

I think the main downside to this approach is probably that the type won't make much sense to users in the non-param case.  But as mentioned in our meeting, I think the interesting question is the degree to which we can hide implementation details from the users so that they don't need to know about them and we can change them.  Confusion about `createVersion(x, y, z).type` or potential changes to what it would give in the future is the main thing I see, but maybe there are others as well?

All that said, I like this direction better than having two types if there aren't gotchas or usability issues that we're not anticipating.

**Comment by vasslitvinov** [comment]
If we are combining the param and non-param flavors into a single type, here is an obvious route:
* start `sourceVersion` with a param field indicating which way it is going,
* have param and non-param fields for each version component that are chosen between based on that first param.

Observe that we still need params for each version component, which are unused for the non-param flavor.

An alternative is to do away the first param that tells which flavor it is -- because it is unnecessary. Instead, use the param for each version component for that purpose. If we have to have those params anyway, may as well use them.

```chpl
record sourceVersion {
  param majorP;
  param minorP;
  // analogously for 'update' and 'commit'

  const majorC: if majorP == none then int else nothing;
  const minorC: if minorP == none then int else nothing;

  proc major param where majorP != none return majorC;
  proc major param where majorP == none return majorC;

  proc minor param where minorP != none return minorC;
  proc minor param where minorP == none return minorC;
}
```

**Comment by lydia-duncan** [comment]
Vass, I think what you are proposing would enable mixing and matching which of the information is param and which isn't, which I think you hinted at in the meeting?

**Comment by vasslitvinov** [comment]
> what you are proposing would enable mixing and matching which of the information is param and which isn't

That's right. However, it is probably not a significant benefit.

**Comment by bradcray** [comment]
> However, it is probably not a significant benefit.

I'd agree with this, since if the whole thing wasn't param, no compile-time comparison would be possible.

> If we are combining the param and non-param flavors into a single type, here is an obvious route:

Vass, I realize that there may have been a race condition between our comments, but it sounds like this is what I sketched out above, is that right?

**Comment by vasslitvinov** [comment]
> Vass, I realize that there may have been a race condition between our comments, but it sounds like this is what I sketched out above, is that right?

Yes, on both counts.

**Comment by vasslitvinov** [comment]
If we are shooting for param versions to become "param records" some time in the future, we should mark explicitly-instantiated `sourceVersion` types, such as `sourceVersion(true, 1, 2, 5)`, as well as direct field accesses, as unstable. This will allow us to trim down `sourceVersion` in the future.

**Comment by lydia-duncan** [comment]
> If we are shooting for param versions to become "param records" some time in the future, we should mark explicitly-instantiated `sourceVersion` types, such as `sourceVersion(true, 1, 2, 5)`, as well as direct field accesses, as unstable. This will allow us to trim down `sourceVersion` in the future.

If the fields aren't going to be documented, I don't think we need to mark them as unstable (and doing so will generate warning messages whenever we use them).  The type description is interesting, though, I'm not sure how we'd accomplish that but agree it does seem like something the user shouldn't be relying on

**Comment by bradcray** [comment]
> The type description is interesting, though, I'm not sure how we'd accomplish that but agree it does seem like something the user shouldn't be relying on

I wonder whether we can just avoid documenting the type in some way in the module documentation, to avoid committing to a specific representation (e.g., similar to what we do for string and bytes, I think?  though this case is slightly weirder given that it's not part of the language).

In any case, let's see if the sketch is tractable and attractive for the current use cases first, and then worry about what aspects expose more of the implementation than we're happy with and consider what to do about them then.

**Comment by arezaii** [comment]
Thanks for all the suggestions and discussion on this! I have made a new draft PR to demonstrate how this could be implemented at https://github.com/chapel-lang/chapel/pull/20398

**Comment by arezaii** [comment]
After wrestling with this for some time, I took a step back and evaluated _why_ we even wanted to add a non-param type to `Version` in the first place. The motivation stemmed from a need in Mason, which has its own form of a non-param `Version`. I hoped it would be a natural use of Chapel's `Version` module, but the paramness of `sourceVersion` made it ill-suited for Mason's needs. 

Looking at the cost-benefit of adding this for 2.0 stabilization, I think it would be best to stabilize `Version` with just the current param type, and continue to let Mason use its custom solution. If we wanted or needed a more robust `Version` package in the future, we could provide more flexibility to the user by looking at the [Rust crate](https://docs.rs/versions/latest/versions/) for inspiration and releasing it as a new package, possibly through Mason.

I'd like to know if other interested parties agree with this decision or not.

**Comment by lydia-duncan** [comment]
I was originally a little worried because it seemed like a real use of the Version module where we fall down, but it sounds like the Mason situation is more a "make the type less of a singleton use case" than a problem with the type as it stands today.  Does that sound right?  If so, I think what you're proposing here is reasonable

**Comment by bradcray** [comment]
> I took a step back and evaluated why we even wanted to add a non-param type to Version in the first place. The motivation stemmed from a need in Mason, which has its own form of a non-param Version. I hoped it would be a natural use of Chapel's Version module, but the paramness of sourceVersion made it ill-suited for Mason's needs.

I'd describe it more as:  "We wanted to use mason's desire for a non-param `version` type as an opportunity to make sure we weren't painting ourselves into a corner in the standard module design, or wouldn't end up regretting the type name or behavior that we currently support."

I think some conclusions we reached were:
* having a non-param version type would be useful
* if we did have one, `sourceVersion` isn't a particularly clear or descriptive name for the current version type
* maybe unifying on a single version type that could be `param` or non- would achieve both of these goals

The exploration of unifying to a single type seemed promising where the main thing that gave me pause (as a distant observer) was the fact that [`createVersion(1, 2, 3)` would necessarily create a `param` version](https://github.com/chapel-lang/chapel/pull/20398#issuecomment-1219533220) when—maybe—I'd just wanted to use that as the initial value, but then re-assign it later.  This seems addressable (e.g., maybe we make two `createVersion()` overloads or have it take a required `param vs. non` flag), but definitely took some wind out of my sails.

The other thing that gave me pause was that Ahmad seemed to lose his faith in the approach, though I'm not sure what he'd (you'd) consider the biggest problems to be.

If we were to decide to throw in the towel on supporting a unified type, I'm less certain why we wouldn't go back to supporting two distinct types in a single module—a param and a non-param version.  Was the only show-stopper there the naming challenge?  Or were there other downsides that I'm forgetting?

**Comment by arezaii** [comment]
After talking with @bradcray about this yesterday, I am more confident that we can implement at two-type solution and arrive at names we all find...tolerable? 

To summarize:
- this came about as a way to remove a custom built type in Mason and replace it with something from the standard library
- we have a two-type draft PR up at https://github.com/chapel-lang/chapel/pull/19300
  - the two types worked well, except we couldn't agree on names for them
  - in module re-review, we decided to try a unified type with a `param` flag to indicate which flavor it was
- we have a unified type draft PR up at https://github.com/chapel-lang/chapel/pull/20398
  - there were [ergonomic challenges with the unified type approach](https://github.com/chapel-lang/chapel/pull/20398#issuecomment-1219533220)

It sounds like people see some value in having a `version` type that can be defined and manipulated at runtime, so let's see if we can agree on some names.

<details>

<summary>
Names we have previously discussed in this thread:
</summary>

| **param type name** | **non-param type name** |
|---------------------|-------------------------|
| sourceVersion       | programVersion          |
| sourceVersion       | runTimeVersion          |
| sourceVersion       | compiledVersion         |
| sourceVersion       | builtVersion            |
| sourceVersion       | artifactVersion         |
| staticVersion       | dynamicVersion          |
| sourceVersion       | packageVersion          |
| compileTimeVersion  | runTimeVersion          |

</details>

New options we haven't discussed:
| param version    | non-param version |
|------------------|-------------------|
| versionValue     | *                  |

*Originally I thought `version` would be a nice fit here, but looking at https://github.com/chapel-lang/chapel/issues/15911, it seems like `version` was dis-favored because of the potential collision with a variable named `version`

I would like to revisit the naming discussion to see if we can agree on some names for the types and the corresponding generator functions for them, e.g., `createVersion`, `createVersionValue`

**Comment by bradcray** [comment]
> Originally I thought version would be a nice fit here, but looking at https://github.com/chapel-lang/chapel/issues/15911, it seems like version was dis-favored because of the potential collision with a variable named version

Oh yeah, I forgot about that reason...  Hmm, do I really believe in it?  In practice, I'm not convinced a lot of code will actually refer to the type name.  It's not like we're reserving a new keyword.  Our module shadowing / use / import rules have improved a lot recently.  And of course, a user can always disambiguate using `Version.version`, or by renaming on `use`/`import`, say. 

I guess I'd be curious what we'd see if we ran a paratest on a branch that used `version` as the type name, as that certainly seems like the most logical/obvious choice.

`paramVersion` also came up in this conversation, and I came up with what I think is a compelling reason not to use it (aside from my general dislike for propagating Chapel's worst term, `param` any further than necessary).

Let’s say that the execution-time / flexible version of the type is just called `version` and that the factory for creating new instances of it is `createVersion(x, y, z)`, such that I could write:

```chapel
const v: version = createVersion(1, 2, 3);
```

In that world, the type phrase `createParamVersion()` suggests to me that it could be used to write something like:

```chapel
param v: version = createParamVersion(1, 2, 3);
```

That is, that it would create a `param` instance of the same `version` type.  But since, in this case, we’re actually using a distinct type with a distinct name (and moreover, one that doesn’t support params), I'd actually have to write something like:

```chapel
const v: paramVersion = createParamVersion();
```

and that seems like a big impedance mismatch to me (i.e., if this thing is really a `param`, why is the declaration a `const` (or `var`)?  And why is the type name different?)  

Whereas if we know we’re calling the type something different, something like:

```chapel
const v: versionValue = createVersionValue();
```

makes that mismatch go away and also describes what the thing is representing "a version value 1.2.3, much as 42 is an integer value" using much more plain terms.  E.g., `if chplVersion < createVersionValue(1, 28, 0) then compilerError("You're out of date");`

**Comment by mppf** [comment]
@arezaii - do you want to make a little poll on the names? You could consider having a different comment for each, so you could update the comment with Pros/Cons if/when people identify them (and link to Brad's comment about `paramVersion`).

**Comment by arezaii** [comment]
React to vote for the non-param type to be named `version`

**Comment by arezaii** [comment]
React to vote for the param type to be named `versionValue`

**Comment by arezaii** [comment]
React to vote for the non-param type to be named something other than `version`

**Comment by bradcray** [comment]
> React to vote for the non-param type to be named version

I like this, but would be curious what the practical effect would be on the current code base, if any.  I'm also curious what type name(s) the Rust package you mentioned liking used.

**Comment by arezaii** [comment]
The `Versioning` crate has three types:

- `SemVer`
- `Version`
- `Mess` 

Of those, `SemVer` is the most restrictive and most closely matches what we our type would look like. For reference, here are the fields for `SemVer`:
```Rust
pub struct SemVer {
    pub major: u32,
    pub minor: u32,
    pub patch: u32,
    pub pre_rel: Option <Release>,
    pub meta: Option<String>,
}
```
*Docs for Versioning Rust crate, for reference: https://docs.rs/versions/latest/versions/enum.Versioning.html


So maybe the 'other' choice would be `semVer` or `semanticVersion` to be verbose. If we went that route, maybe `semVerValue` would be a more precise type name for the param version.

**Comment by bradcray** [comment]
To me, the relevant thing is that they have a type named version even if it's not an exact match with ours, so that makes me feel OK about using that name.  I'm also open to `semanticVersion`, though with the SHA, it's also a bit more than that (so doesn't feel like a precise description, at least based on my memory of how semantic versions are defined).

**Comment by bradcray** [comment]
Hmm, maybe I was misremembering about semantic versioning not having a concept of a SHA.  From https://semver.org/#spec-item-9:

> A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version. Identifiers MUST comprise only ASCII alphanumerics and hyphens [0-9A-Za-z-]. Identifiers MUST NOT be empty. Numeric identifiers MUST NOT include leading zeroes. Pre-release versions have a lower precedence than the associated normal version. A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version. Examples: 1.0.0-alpha, 1.0.0-alpha.1, 1.0.0-0.3.7, 1.0.0-x.7.z.92, 1.0.0-x-y-z.–.
> 
> Build metadata MAY be denoted by appending a plus sign and a series of dot separated identifiers immediately following the patch or pre-release version. Identifiers MUST comprise only ASCII alphanumerics and hyphens [0-9A-Za-z-]. Identifiers MUST NOT be empty. Build metadata MUST be ignored when determining version precedence. Thus two versions that differ only in the build metadata, have the same precedence. Examples: 1.0.0-alpha+001, 1.0.0+20130313144700, 1.0.0-beta+exp.sha.5114f85, 1.0.0+21AF26D3—-117B344092BD.

So maybe, with a careful read of this doc, we could get to the point where we felt we had defined a `semanticVersion` type.  That said, I would also be completely fine with using the name `version` and simply saying "versions defined by the Version module are semantic versions." or "... are modeled after semantic versions."  Tradeoffs in those two approaches, to me, are accuracy in terminology (where the degree to which we do, or want to, conform could be an issue) vs. keeping the name short and sweet vs. reducing the potential for overlap with a variable named `version`.  We could also use `version` in the short term for our loose notion of version and then introduce a more precise `semanticVersion` type later.

**Comment by arezaii** [comment]
At module review today, we decided the following:

* use the name `version` for the non-param type
* the param-type will be called `versionValue`
* we will drop the factory functions and rely on initializers, e.g., `new version(1,2,3)` or `new versionValue(1,2,3)`

votes:
***Name of non-param type***
- value: 7
- semanticVersion: 1
- semVer: 0
- semVersion: 1

***How to identify param type***
- append `value`: 8
- prepend `param`: 0

***How to build a new object of each type: <non-param, param>***
- free factories `createVersion`, `createVersionValue`: 1
- type methods `version.create`, `versionValue.create`: 0
- initializers: `new version`, `new versionValue`: 8

**Comment by arezaii** [comment]
The types `version` and `versionValue` were added in https://github.com/chapel-lang/chapel/pull/19300 to handle non-param and param cases, respectively.

https://github.com/chapel-lang/chapel/pull/21088 reinstates and marks as deprecated the old `createVersion` function and the `sourceVersion` type has been marked as deprecated and made an alias for `versionValue`.

**Comment by arezaii** [comment]
Closing as this was implemented in #19300 and #21088
